sql/analyzer/resolve_ctes.go

```diff
@@ -82,9 +84,40 @@ func resolveCtesInNode(ctx *sql.Context, a *Analyzer, node sql.Node, scope *Scop
 			switch n := node.(type) {
 			case *plan.UnresolvedTable:
 				lowerName := strings.ToLower(n.Name())
-				if ctes[lowerName] != nil {
-					return ctes[lowerName], transform.NewTree, nil
+				cte := ctes[lowerName]
+				if cte == nil {
+					return n, transform.SameTree, nil
+				}
+
+				if isRecursive {
+					return cte, transform.NewTree, nil
+				}
+
+				// look for self references in non-recursive cte
+				subquery, ok := cte.(*plan.SubqueryAlias)
```

Comment:
explicitly cast, we don't want to accidentally ship code that doesn't panic here